### DISCUSSION TITLE
Problem setting up NTP Time Sync

### DISCUSSION THREAD

**Discussion by ldb20151** [aaronwmorris/indi-allsky#1700] [Q&A] [answered]
After working through my GPS initialization issues in 1698, I convinced myself I was ready to proceed with installation of ntpsec.  Installation appeared to proceed with no issues.  I saved and have reviewed the log -- all looks OK from my perspective.  
When I proceed to configure:  /etc/ntpsec/ntp.conf I receive a "permission denied" error.  Just to see where the system was I went ahead and tried to validate the time sync setup with  ntpq -p  I received a complex printout -- not sure how to interpret it.  It appeared to show time relative to other internet sources but I can't judge good vs bad with  what I know.  
Questions:  1) How do I interpret the validation info to tell if all is working correctly?  If "remote" and "refid" columns refer to compared sources, wouldn't I expect to see GPS as one of these sources?  I also don't see any reference to 127 in either column, though there do appear to be .POOL. and other sources listed. I'm inserting below as best I can -- sorry for the poor formatting.
     remote           refid      st t when poll reach   delay   offset   jitter
===============================================================================
 0.debian.pool.n .POOL.          16 p    -  256    0   0.0000   0.0000   0.0001
 1.debian.pool.n .POOL.          16 p    -   64    0   0.0000   0.0000   0.0001
 2.debian.pool.n .POOL.          16 p    -   64    0   0.0000   0.0000   0.0001
 3.debian.pool.n .POOL.          16 p    -   64    0   0.0000   0.0000   0.0001
#time3.lshiy.com 218.73.139.35    2 u   39   64  377 212.1545  17.2951  42.3735
-t1.time.gq1.yah 98.137.249.214   2 u   49   64  377  77.4410  26.5090  41.4023
+23.150.40.242   204.9.54.119     2 u    8   64  167  46.0108  19.3603  28.3449
+coco.presumed.n 173.73.96.68     2 u   10   64  377  10.0497  53.6671  27.5131
-144.202.41.38.v 23.150.41.122    5 u   42   64  373  34.2284  58.9533  36.6977
+66.205.249.28   204.9.54.119     2 u   43   64  177  27.2700  54.9055  33.7440
#104-167-215-195 128.138.140.44   2 u  178   64   64  43.3153 -50.8364  58.5879
#static-72-87-88 185.125.190.56   3 u   36   64  377  25.2096  59.1146  34.3922
+time2.tritan.ho 169.117.81.12    2 u   28   64  321  39.6482 -50.2633  66.1864
#connected.by.fr 10.1.105.4       2 u   32   64  377 100.7847  58.9414  30.3287
+mailhost.netads 206.108.0.131    2 u   39   64  377  15.7032  17.0581  40.9586
-ntp.speculation 128.9.176.30     2 u   42   64  377  70.4508  73.4844  50.8602
-h96-60-160-227. .PPS.            1 u   35   64  357  87.2635   6.7489  38.6811
*t1.time.bf1.yah 98.139.133.62    2 u   27   64  377  18.7616  17.5688  32.0928
+sensei.ruselabs 128.138.140.44   2 u   37   64  377  68.3962  14.4671  27.7650
#50.218.103.254  132.163.96.1     2 u  149   64   64  21.1707  18.0302  67.8268

2) Why did I get permission denied when trying to configure?  Since I'm internet connected, do I need to worry about configuration?
As always, help is appreciated.
-Larry

**Comment by aaronwmorris** [comment]
You will need to edit the ntp config file via `sudo`.  The output above just means you are ntp syncing using ntp hosts on the Internet, which is expected.  Once you add the configuration in the GPS wiki page, you should see something like this:

```
$ ntpq -p
     remote                                   refid      st t when poll reach   delay   offset   jitter
=======================================================================================================
xSHM(0)                                  .GPS.            1 l   14   16  377   0.0000 -67.2402   1.5992
 0.ubuntu.pool.ntp.org                   .POOL.          16 p    -   64    0   0.0000   0.0000   0.0005
 1.ubuntu.pool.ntp.org                   .POOL.          16 p    -   64    0   0.0000   0.0000   0.0005
 2.ubuntu.pool.ntp.org                   .POOL.          16 p    -  256    0   0.0000   0.0000   0.0005
 3.ubuntu.pool.ntp.org                   .POOL.          16 p    -  256    0   0.0000   0.0000   0.0005
-alphyn.canonical.com                    132.163.96.1     2 u   97 1024  377  50.1493   7.4558   1.8710
+149.28.61.105.vultrusercontent.com      208.113.130.146  3 u  394 1024  377  43.2707   6.7218   1.7831
*99-28-14-242.lightspeed.iplsin.sbcgloba ....             1 u  563 1024  375  50.5627   1.4358   3.8964
+time.tritan.host                        169.117.81.12    2 u  304 1024  377  39.0295   6.4774   2.1746
-pool-173-71-68-71.cmdnnj.fios.verizon.n .PPS.            1 u  248 1024  377  53.3408  11.5682   3.6370
-ntp4.glypnod.com                        129.6.15.27      2 u  431 1024  377  60.4248   8.4087   1.1600
```

**Reply by ldb20151** [reply]
I edited the config file as described in the wiki.  The first time the file was completely blank and that surprised me.  I added the new text and things didn't work correctly so I went back to check my entries and noticed the file had significant config text that I didn't add -- my entries were gone.  Not sure what I did there but I corrected the ntp.conf file per the wiki, restarted the NTP server and now I'm getting the following output from ntpq -p:  
REMOTE         Refid             st  t     when   poll    reach     delay          offset           jitter   
xSHM(0)          .GPS.            1   l       1        16     377        0.0000       -655.277      20.9238
Everything looks good to me except offset and jitter appear larger than I'd expect.  I guess this may be due to the fact that the antenna is with me here inside the workshop.  Just can't reach a clear view of sky and too cold to work outside right now.  Still, this seems like a pretty big offset and I'm concerned ntp isn't sync'ing properly.  FYI I did not "uncomment" the PPE lines in the ntp.conf file so that feature is not enabled.  Do I need that level of precision for the camera and would it help the delay numbers?  Just a thought.
I still can't get gnome or scrot to work on this system, so again I'm sorry for the lack of a screen shot.  The other parameters are in line with the GPS but  2+ orders of magnitude smaller delay values.  Next largest jitter is about 10.7.
Is this good to go so I can move to next steps or are there other checks I should do first?

**Reply by aaronwmorris** [reply, marked as answer]
You definitely do not need the PPS signal.  That is only really useful for people who need high level precision for occlusions or eclipses, etc.


It is odd that `/etc/ntpsec/ntp.conf` was blank.  This what my system with GPS looks like (excluding comments):

```
driftfile /var/lib/ntpsec/ntp.drift
leapfile /usr/share/zoneinfo/leap-seconds.list

tos maxclock 11

### I only added the following two lines
server 127.127.28.0 minpoll 4 maxpoll 4 prefer
fudge 127.127.28.0 refid GPS stratum 1
###

pool 0.ubuntu.pool.ntp.org iburst
pool 1.ubuntu.pool.ntp.org iburst
pool 2.ubuntu.pool.ntp.org iburst
pool 3.ubuntu.pool.ntp.org iburst

server ntp.ubuntu.com

restrict default kod nomodify nopeer noquery limited

restrict 127.0.0.1
restrict ::1
```

**Reply by ldb20151** [reply]
Thank you for the ntp.conf example.  The bottom line, I think I solved the time sync issue (super high delay/jitter).  I'm still not able to verify -- don't know how -- AllSky is using the data.  Maybe these checks are all that's needed?

The details....
First, regarding the mysterious empty /etc/ntpsec/ntp.conf file at my first edit.  You're correct, that was entirely my mistake.  I had a typo the first time I attempted to edit the file.  To my surprise, instead of throwing an error and telling me that my fat-fingered file name didn't exist, the nano editor just opened a new, empty file.  When I attempted to "re-open" it later to double check my entries, I typed the correct file name, saw the needed edits were missing ('cause this was the first time I had accessed the correct file), added them and things started working, albeit with the higher than expected time delay.  At this point, I still have an unneeded "typo file" at the same location as the correct ntp.conf file.  I don't know how to delete it so for now I'm ignoring it.  Is this going to hurt anything?  How can I delete it without messing up other files in this area?  So far, it doesn't seem to be having an adverse effect.

Regarding the delay, for reasons I can't explain, the ntp.conf text was missing the line "server ntp.ubuntu.com" -- in my case "server ntp.debian.com", I think since those are the default pool references.  Seems they should be the same so I made the adjustment and its working.  There was also an extra line I don't recall, so I removed it to match your file.  Once I adjusted the ntp.conf file to match yours (except for the debian reference) and reset the NTP, my GPS time fell nicely in line with your example.  

I have not yet checked to see if AllSky is actually using the GPS time and locations data -- is there a way to verify that other than ntpq -p?  
Thx much!

**Reply by aaronwmorris** [reply]
`ntpq -p` is really the only method of validating that I am aware of.

**Reply by ldb20151** [reply]
I've gone through everything and finally have ntpq -p producing the correct response.  Once in the camera application and signed in to indi-allsky, I went to the systems/time tab and see the following:
NTP Available
False
NTP Enabled
False
NTP Synchronized
True
RTC Available
False
Given I'm using onboard GPS, shouldn't all of these read "false"?  I've gone back over the steps to install/enable GPS and can't find any issues.  We are having very heavy rain today.  The GPS is dropping out once in a while but says it's locked now.  Am I understanding this correctly?  If so, do you have any suggestions on where I should look to resolve the problem?  
I must be getting close.  I sure hate to give up on it now but this has proven to be harder than I expected.

**Reply by aaronwmorris** [reply]
This is kind of to be expected.  These indicators are from systemd and you are bypassing the systemd timesync configs.  I am surprised it detected the system was NTP synchronized.
